src/interpreter/mod.rs

```diff
@@ -734,8 +734,17 @@ impl<'a, 'tcx> EvalContext<'a, 'tcx> {
             ty::TyRef(_, ty::TypeAndMut { ty, .. }) |
             ty::TyRawPtr(ty::TypeAndMut { ty, .. }) |
             ty::TyBox(ty) => {
-                assert_eq!(field_index, 0);
-                Ok(ty)
+                if self.type_is_sized(ty) {
+                    assert_eq!(field_index, 0);
+                    Ok(self.tcx.mk_imm_ptr(self.tcx.types.never))
+                } else {
+                    match (field_index, &ty.sty) {
+                        (1, &ty::TySlice(_)) => Ok(self.tcx.types.usize),
+                        (1, &ty::TyTrait(_)) |
+                        (0, _) => Ok(self.tcx.mk_imm_ptr(self.tcx.types.never)),
```

Comment:
I'd use a pointer to `u8` to avoid any other implications. Btw you need to use `struct_tail` to get `ty`.
Also I don't think the sized case above makes sense? We only treat a fat pointer like a pair.